### ISSUE TITLE
Control issue with latest Nespad changes & WS PiZero

### ISSUE THREAD

**Issue by DynaMight1124** [fhoedemakers/pico_shared#8]
Hi

Thanks for the continued improvements, just updated my PicoNES's to 0.23, both of the PCB versions (0.1 & 2.0) worked fine, however when I updated my WS PiZero, the controls didnt work, well not properly. I could press start if I plugged into player 2 but couldnt use any direction buttons. I am using an aftermarket NES controller (original connectors etc, not USB) I havent tested USB controllers but happy to test if that helps. 

I flashed back to 0.22 and the controls are working fine. I assume it is related to the latest update with the CLK high/low. Not sure why the WS board is different, the pins of the NES connector are connected directly to the GPIO on the WS board.

**Comment by fhoedemakers** [comment]
Hi @ManCloud, maybe you can be of any help. We are experiencing some issues with the Nespad changes when using the WaveShare Rp2040-PiZero. No other config has these issues.  Since you are planning to change the board for WaveShare support it may also be of concern for you:

These are the issues we experience:

- Not working at all, or only some buttons using aftermarket controllers (with original connectors)
- Erratic behaviour using original controllers. Like pressing down/up in the menu is skipping multiple lines, in-game erratic movement of the characters you control.

I'll be glad to be of any assistance in this.

Thanks.

**Comment by ManCloud** [comment]
I just ordered a board to test. Shall arrive on Monday 👍

**Comment by fhoedemakers** [comment]
Thanks!

**Comment by ManCloud** [comment]
I had some time to check differences: The WS PiZero has a 2.2k pull-down on NES_DATA.
This might interfer with timings or levels. 

This is the only difference that i noticed until now. 
I'll try to "simulate" that by adding a 2.2k Resistor to my other boards and measure the difference.

if it's really the resistor a quick fix would be to "simply" remove the resistor R34.

**Comment by fhoedemakers** [comment]
@ManCloud  Merged some changes into the pico_shared main branch. Make sure to sync  main of this repository with your fork. For the time being, i have changed nespad.cpp to use the old code for the WaveShare config. All the other configs still use the new code.

For this there is also made a change in CMakeLists.txt in the pico-infoNesPlus repo. I added HW_CONFIG to the compile definitions.

![image](https://github.com/user-attachments/assets/0d453f26-5d14-4af3-aa3e-b7a8469489e5)

Happy New Year!

**Comment by ManCloud** [comment]
I now finally checked the WS PiZero and unfortunately i can't reproduce the issue.
My Controller works just fine with FW 0.23.

The reason it works for me might in fact be, that i use OEM controllers. Unfortunately i don't have other controllers to test.
Controller is detected perfectly in P1 as well as P2.

**Comment by fhoedemakers** [comment]
Hi, Thanks for testing, much appreaciated.

Maybe it is related to a specific HW revision of the board.
For now, the WaveShare binary will be released with the old controller code. All the other binaries will use the new code.
I recently got some third party controllers from AliExpress. I will test them as soon as my daytime job permits.

Is it an idea i send you the board along with a controller when i am able to reproduce the issue?

Thanks!

**Comment by ManCloud** [comment]
Yes i guess if you're able to recreate the issue this would help 👍

**Comment by fhoedemakers** [comment]
Hi,
I can reproduce the issue on V0.23 with my board:
- Using the official OEM controller, it behaves erratically
- Using an AliExpress third party controller, only the A-button seems to work

Both controllers work fine with the original nespad code.

I will send you my board and the AlieExpress controller. Can you DM me your Address via Discord private message.

**Comment by DynaMight1124** [comment]
Hi, all good thanks. I've flashed the new version and its now working perfectly with my aftermarket NES controller. Tested multiple games over 15-20mins and seemed to be perfect. 

Thank you to you and ManCloud

On a slightly different note, I was ordering some PCB's from JLCPCB the other day and decided to order some SLA versions of my shell. It should look very clean, there isnt a load of colour options so tried to match the best I could based on their descriptions.

**Comment by fhoedemakers** [comment]
Thanks for testing! Appreciated! Will post a new release of the NES and Master System emulators very soon.
Curious how the shells turn out!

**Comment by DynaMight1124** [comment]
The case arrived. Looks pretty nice and very well priced from JLC. You can see the layer lines a little but its nicely finished. Kinda wished I took the risk and ordered one clear, I think it would had looked pretty cool.

![Image](https://github.com/user-attachments/assets/82e61dec-0fa1-40b2-9e74-1cc49b784ae2)

**Comment by fhoedemakers** [comment]
Looks good! Transparant would be cool.  😄
